### ISSUE TITLE
Dynamic @gen functions do not perform type-checks

### ISSUE THREAD

**Issue by ztangent** [probcomp/Gen.jl#224]
Currently, the DML does not pass along type information to the internal `julia_fn`, allowing you to do stuff like:
```julia
julia> @gen foo(x::Bool) = x;
julia> foo(5)
5
```
This could easily be addressed by modifying
```julia
function arg_to_ast(arg::Argument)
    ast = esc(arg.name)
    if (arg.default != nothing)
        default = something(arg.default)
        ast = Expr(:kw, ast, esc(default))
    end
    ast
end
```
defined [here](https://github.com/probcomp/Gen/blob/master/src/dsl/dynamic.jl#L3-L10) to construct an AST that contains the type annotation. However, the last time I tried to do this (as part of the changes in #195), this led to errors with some of the autodiff unit tests, because apparently the current autodiff library we're using wraps Floats in some wrapper type. I'm not sure how to address this -- I suppose it might be good to wait until the AD refactor suggested in #128 happens.

**Comment by ztangent** [comment]
An example error message from when I made that change last time:
```julia
Got exception outside of a @test
  MethodError: no method matching (::var"##foo#445#17"{DynamicDSLFunction{Any}})(::Gen.GFBackpropTraceState, ::ReverseDiff.TrackedReal{Float64,Float64,Nothing})
  Closest candidates are:
    Any(::Any, !Matched::Float64) at C:\Users\...\Programs\Gen\test\dynamic_dsl.jl:281
```

**Comment by marcoct** [comment]
@ztangent I think that the general way to add support for this without breaking AD (and more generally reducing interactions between various non-standard interpretations) is to generate multiple versions of the Julia function, that are used for different purposes internally. For example, AD support based on ReverseDiff.jl (which relies on the function having abstract argument types) could use a version of the Julia function that does not include the types given in the function signature, but generate, update and other methods like those would use a version that includes the types in the function signature. AD based on Zygote.jl will use the same typed version of the function as generate, etc.

As an example of an alternative, I don't think it makes sense to let details of how ReverseDiff leak into the modeling language itself (for example, by requiring that the user make use of abstract types like `Real` and not concrete types like `Float64` in their generative function argument signature).

For some context, the dynamic modeling language is distinguished by the fact that it implements the GFI using  non-standard end-to-end interpretations of the Julia function body that can include running arbitrary Julia code. Currently, the same Julia function is interpreted different ways using effect handlers, but we can also have distinct Julia functions when necessary  (so there is a sort of two-layer hierarchy in the implementation).

**Comment by ztangent** [comment]
Addressed by #314.
